### ISSUE TITLE
Incorrect autocorrection by Style/CollectionCompact

### ISSUE THREAD

**Issue by wendelscardua** [rubocop-hq/rubocop#9023]
The new `Style/CollectionCompact` cop may change some code behavior. It assumes that `reject` can be replaced with `compact` even when it shouldn't. Probably it is applying to an `Array#reject` a correction that would only be valid for a `Hash#reject`. See below for an example.

--------

## Expected behavior
When a `reject` method is called for an array with a block with two arguments, and this block only checks if the second argument is nil, this `reject` shouldn't be replaced with a `compact`. 

## Actual behavior
`reject` gets replaced with `compact` in that scenario.

## Steps to reproduce the problem

* Create this ruby file (lets call it `test.rb`):
```ruby
x = [[1, nil], [2, 3], [4, nil]]
x = x.reject { |_a, b| b.nil? }
puts x.size
```
* Run it
```sh
$ ruby test.rb
1
```
* Autocorrect this file with rubocop, selecting only the Style/CollectionCompact cop:
```sh
$ bundle exec rubocop --only Style/CollectionCompact test.rb -a
(some lines about new cops not enabled omited)
Inspecting 1 file
Scanning /home/wendel/projects/voltera-api/test.rb
C

Offenses:

test.rb:2:7: C: [Corrected] Style/CollectionCompact: Use compact instead of reject { |_x, y| y.nil? }.
x = x.reject { |_x, y| y.nil? }
      ^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense corrected
```
* File content is now:
```ruby
x = [[1, 2], [3, nil], [4, nil]]
x = x.compact
puts x.size
```
* Run the modified script:
```sh
$ ruby test.rb
3
```
* The output is now 3 instead of 1

## RuboCop version

```
$ bundle exec rubocop -V
1.2.0 (using Parser 2.7.2.0, rubocop-ast 1.1.1, running on ruby 2.7.1 x86_64-linux)
  - rubocop-performance 1.8.1
  - rubocop-rails 2.8.1
  - rubocop-rake 0.5.1
```

**Comment by koic** [comment]
Thank you for detailed feedback. I think this is difficult to detect without false positives and I opened a PR #9024.
